backup and restore docs & api (#364)
upbound/docs: content/spaces/backup-and-restore.md

---

> title: Backup and restore Upbound Spaces

Reviewer: ```suggestion
title: Backup and restore
```

The title as-is makes it sound like it concerns backup/restore of the Space itself, when its really about the control planes in a Space.

weight: 120
description: Enable and manage backups in your Upbound Space.
---

{{< hint "important" >}}

> This functionality is in preview and requires Spaces `v1.3.0`.

Reviewer: ```suggestion
This feature is in preview, requires Spaces `v1.3.0`, and is off by default. To enable, set `features.alpha.sharedBackup.enabled=true` when installing Spaces:

```bash
up space init --token-file="${SPACES_TOKEN_PATH}" "v${SPACES_VERSION}" \
  ...
  --set "features.alpha.sharedBackup.enabled=true" 
 ```
```

{{< /hint >}}

> Upbound allows you to configure a backup and restore process. This feature helps manage your important control plane data and ensures security and availability best practices.

Reviewer: ```suggestion
Upbound allows you to configure backup and restore for control planes in the Space. This feature helps manage your important control plane data and ensures security and availability best practices.
```

## Usage

### Backup

#### Create a backup secret

> Before you begin the backup process, you need to create a secret for your Kubernetes configuration to allow for communication between your backup and your restored configuration.

Reviewer: ```suggestion
Before you can configure backup schedules and initiate manual backups, you need to create a secret containing auth credentials to allow for communication between the Space and a storage target, such as AWS S3.
```

```bash
KUBECONFIG=/tmp/ctp.yaml kubectl create secret generic super-secret-secret -n default --from-literal=password=supersecret
```

> #### Single backup

Reviewer: I think the flow should be to first create a `SharedBackupConfig`, then make a schedule, then discuss how to do a manual backup, then do a restore

      source: Secret
      secretRef:
        name: bucket-creds
        key: creds
---
apiVersion: v1
kind: Secret
metadata:
  name: bucket-creds
  namespace: default
stringData:
  creds: |
    [default]
    aws_access_key_id=***************
    aws_secret_access_key=*************************
```

Then create a `ControlPlane` configuration with that secret reference:

```yaml
apiVersion: spaces.upbound.io/v1beta1
kind: ControlPlane
metadata:
  labels:
    org: foo
  name: my-awesome-ctp
  namespace: default
spec:
  backup:
    sharedBackupConfigRef:
      name: default
  writeConnectionSecretToRef:
    name: kubeconfig-my-awesome-ctp
---
apiVersion: spaces.upbound.io/v1beta1
kind: ControlPlane
metadata:
  labels:
    org: foo
  name: my-second-awesome-ctp
  namespace: default
spec:
  backup:
    sharedBackupConfigRef:
      name: default
  writeConnectionSecretToRef:
    name: kubeconfig-my-second-awesome-ctp
```

Then a `SharedBackup` configuration:

```yaml
apiVersion: spaces.upbound.io/v1alpha1
kind: SharedBackup
metadata:
  name: custom-shared-backup
  namespace: default
spec:
  controlPlaneSelector:
    labelSelectors:
    - matchLabels:
        org: foo
```

> This example configuration results in both control planes backed up and uploaded to the S3 endpoint.

Reviewer: ```suggestion
This example results in both control planes backed up and uploaded to the S3 endpoint.
```

### Create a scheduled backup

> Instead of a one-time backup with the `SharedBackup` kind, you can create a `SharedBackupSchedule` configuration:

Reviewer: ```suggestion
Instead of a one-time backup with the `SharedBackup` kind, you can create a `SharedBackupSchedule:
```

instead of calling these `configurations` (heavily overloaded term in our space due to Crossplane Configurations), what do you think about referring to these as resources, objects, or just dropping configuration entirely? Think it reads just as fine.

```yaml
apiVersion: spaces.upbound.io/v1alpha1
kind: SharedBackupSchedule
metadata:
  name: custom-schedule
  namespace: default
spec: "@every 1m"
  schedule:
  controlPlaneSelector:
    labelSelectors:
    - matchLabels:
        org: foo
```

> Both control planes with the matching labels will be backed up every minute.

Reviewer: ```suggestion
Both control planes with the matching labels will be backed up every hour.
```

Pairs with my other suggestion

### Restore

To restore from a backup, check the `ControlPlane` to make sure it is ready.

```bash
kubectl wait controlplane my-awesome-ctp --for condition=Ready=True --timeout=3600s && \
kubectl get secret kubeconfig-my-awesome-ctp -n default -o jsonpath='{.data.kubeconfig}' | base64 -d > /tmp/ctp.yaml
```

Next, initiate the restore process from the backup you created previously.

```yaml
---
apiVersion: spaces.upbound.io/v1beta1
kind: ControlPlane
metadata:
  name: my-awesome-restored-ctp
  namespace: default
spec:
  restore:
    source:
      apiGroup: spaces.upbound.io
      kind: Backup
      name: my-awesome-ctp-backup
  backup:
    sharedBackupConfigRef:
      name: default
  writeConnectionSecretToRef:
    name: kubeconfig-my-awesome-restored-ctp
```

## Considerations

> - Deleting the `SharedBackup` and `SharedBackupSchedule` configurations does not automatically delete the created backups, unless `useOwnerReferencesInBackup` is set to true.

Reviewer: ```suggestion
- Deleting the `SharedBackup` and `SharedBackupSchedule` resources don't automatically delete the created backups, unless `useOwnerReferencesInBackup` is set to true.
```
